### PR TITLE
ENH: Add some conversions for DICOM values

### PR DESCRIPTION
**Pull request by moloney**
Adds conversions between DICOM 'TM' value representation and
seconds past midnight, as well as between 'AS' value representation
and years of age.

### CODE REVIEW

File: nibabel/nicom/utils.py [hunk 1]
```diff
@@ -51,3 +53,138 @@ def find_private_section(dcm_data, group_no, creator):
             if creator == name:
                 return elno * 0x100
     return None
+
+
+def tm_to_seconds(time_str):
+    '''Convert a DICOM time value (value representation of 'TM') to the number
```
Comment (matthew-brett) [inline nibabel/nicom/utils.py]: Needs to be single descriptive line - maybe `Convert DICOM time value (VR of 'TM') to seconds past midnight` ?

File: nibabel/nicom/utils.py [hunk 2]
```diff
@@ -51,3 +53,138 @@ def find_private_section(dcm_data, group_no, creator):
...
+    Parameters
+    ----------
+    time_str : str
+        The string value from the DICOM element
+
+    Returns
+    -------
+    sec_past_midnight : float
+        The number of seconds past midnight
+    '''
```
Comment (matthew-brett) [inline nibabel/nicom/utils.py]: Have you got a good reference for the rules you're using below?

File: nibabel/nicom/utils.py [hunk 3]
```diff
@@ -51,3 +53,138 @@ def find_private_section(dcm_data, group_no, creator):
...
+    else:
+        # If there are more than six chars, the seventh position should be
+        # a decimal followed by at least one digit
+        if str_len == 7 or time_str[6] != '.':
+            is_valid = False
+    if not is_valid:
+        raise ValueError("Invalid number of digits for 'TM' VR")
+
+    # Make sure we don't have leading white space
+    if time_str[0] in string.whitespace:
```
Comment (matthew-brett) [inline nibabel/nicom/utils.py]: How about more whitespace?  Maybe `time_str != time_str.lstrip()` ?

File: nibabel/nicom/utils.py [hunk 4]
```diff
@@ -51,3 +53,138 @@ def find_private_section(dcm_data, group_no, creator):
...
+
+    # Allow ACR/NEMA style format which includes colons between hours/minutes
+    # and minutes/seconds
+    colons = [x.start() for x in re.finditer(':', time_str)]
+    if len(colons) > 0:
+        if colons not in ([2], [2, 5]):
+            raise ValueError("Invalid use of colons in 'TM' VR")
+        time_str = time_str.replace(':', '')
+
+    # Make sure the string length is valid
```
Comment (matthew-brett) [inline nibabel/nicom/utils.py]: Would this also work?

```
    time_exp = re.compile(r"^(\d\d)(\d\d)?(\d\d)?(\.\d+)?$")
    match = time_exp.match(time_str)
    if match is None:
        raise ValueError('Invalid tm string "{0}"'.format(time_str))
    parts = [0 if v is None else float(v) for v in match.groups()]
    return parts[0] * 3600 + parts[1] * 60 + parts[2] + parts[3]
```

File: nibabel/nicom/utils.py [hunk 5]
```diff
@@ -51,3 +53,138 @@ def find_private_section(dcm_data, group_no, creator):
...
+    ----------
+    seconds : float
+        Number of seconds past midnights
+
+    Returns
+    -------
+    tm : str
+        String suitable for use as value in DICOM element with VR of 'TM'
+    '''
+    hours = seconds // 3600
```
Comment (matthew-brett) [inline nibabel/nicom/utils.py]: Could use divmod?

File: nibabel/nicom/tests/test_utils.py [hunk 6]
```diff
@@ -67,3 +68,34 @@ def test_find_private_section_real():
...
+
+def test_tm_to_seconds():
+    for str_val in ('', '1', '111', '11111', '111111.', '1111111', '1:11',
+                    ' 111'):
+        assert_raises(ValueError, tm_to_seconds, str_val)
+    assert_almost_equal(tm_to_seconds('01'), 60*60)
+    assert_almost_equal(tm_to_seconds('0101'), 61*60)
+    assert_almost_equal(tm_to_seconds('010101'), 61*60 + 1)
+    assert_almost_equal(tm_to_seconds('010101.001'), 61*60 + 1.001)
+    assert_almost_equal(tm_to_seconds('01:01:01.001'), 61*60 + 1.001)
```
Comment (matthew-brett) [inline nibabel/nicom/tests/test_utils.py]: Add single colon case?

File: nibabel/nicom/utils.py [hunk 7]
```diff
@@ -51,3 +53,138 @@ def find_private_section(dcm_data, group_no, creator):
...
+    return float(result)
+
+
+def seconds_to_tm(seconds):
+    '''Convert a float representing seconds past midnight into DICOM TM value
+
+    Parameters
+    ----------
+    seconds : float
+        Number of seconds past midnights
```
Comment (matthew-brett) [inline nibabel/nicom/utils.py]: Typo midnight

File: nibabel/nicom/utils.py [hunk 8]
```diff
@@ -51,3 +53,138 @@ def find_private_section(dcm_data, group_no, creator):
...
+    age_str : str
+        The string value from the DICOM element
+
+    Returns
+    -------
+    age : float
+        The age of the subject in years
+    '''
+    age_str = age_str.strip()
+    if age_str[-1] == 'Y':
```
Comment (matthew-brett) [inline nibabel/nicom/utils.py]: Could use conversion dictionary, like `CONVERSIONS = {'Y': 1, 'M': 12, 'W': 365. / 7, 'D': 365}` ?

File: nibabel/nicom/utils.py [hunk 9]
```diff
@@ -51,3 +53,138 @@ def find_private_section(dcm_data, group_no, creator):
...
+    minutes = seconds // 60
+    seconds -= minutes * 60
+    res = '%02d%02d%08.5f' % (hours, minutes, seconds)
+    return res
+
+
+def as_to_years(age_str):
+    '''Convert a DICOM age value (value representation of 'AS') to the age in
+    years.
+
```
Comment (matthew-brett) [inline nibabel/nicom/utils.py]: Add reference to http://dicom.nema.org/Dicom/2011/11_05pu.pdf "age string" ?

File: nibabel/nicom/utils.py [hunk 10]
```diff
@@ -51,3 +53,138 @@ def find_private_section(dcm_data, group_no, creator):
...
+    years : float
+        The years of age
+
+    Returns
+    -------
+    as : str
+        String suitable for use as value in DICOM element with VR of 'AS'
+    '''
+    if years == round(years):
+        return '%dY' % years
```
Comment (matthew-brett) [inline nibabel/nicom/utils.py]: String doesn't have to be zero padded, length 4?

File: nibabel/nicom/tests/test_utils.py [hunk 11]
```diff
@@ -67,3 +68,34 @@ def test_find_private_section_real():
...
+
+
+def test_tm_rt():
+    for tm_val in ('010101.00000', '010101.00100', '122432.12345'):
+        assert_equal(tm_val, seconds_to_tm(tm_to_seconds(tm_val)))
+
+
+def test_as_to_years():
+    assert_equal(as_to_years('1'), 1.0)
```
Comment (matthew-brett) [inline nibabel/nicom/tests/test_utils.py]: Is just `1` or `2` valid?  Does it reliably refer to a year?

Do strings have to be zero padded to the left (as in `001Y`)?  Standard seems to suggest that - but maybe not so in practice?

Maybe worth testing strings like `061Y` and `002W`.

**Comment by matthew-brett** [comment]
Brendan - anything I can do to help keep up the momentum?

**Comment by effigies** [comment]
@moloney Any interest in trying to finish these (including #416, #417 and #418 as well) before a new release?

**Comment by ZviBaratz** [comment]
Implemented in dicom_parser, see [Value Representation](https://dicom-parser.readthedocs.io/en/latest/value_representation.html) (particularly the [Age String](https://dicom-parser.readthedocs.io/en/latest/value_representation.html#age-string) example and [`Time`](https://dicom-parser.readthedocs.io/en/latest/modules/dicom_parser.data_elements.html#dicom_parser.data_elements.time.Time) class).

**Comment by moloney** [comment]
Yes, I will close this one as we are moving to use dicom_parser for this functionality.

**Pull request closed without merging**